File: Top_100_Liked_Questions/21._Merge_Two_Sorted_Lists/solution.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var mergeTwoLists = function(l1, l2) {
    let dummyNode = new ListNode();
    let currNode = dummyNode;
    
    while (l1 && l2) {
        if (l1.val < l2.val) {
            currNode.next = l1;
            l1 = l1.next;
        } else {
            currNode.next = l2;
            l2 = l2.next;
        }
        currNode = currNode.next;
    }
    
    currNode.next = l1 ? l1 : l2;
    
    const newHead = dummyNode.next;
    // remove the reference to the dummyNode instance so that the GC removes it
    dummyNode = undefined;
    
    return newHead;
};

File: Top_100_Liked_Questions/21._Merge_Two_Sorted_Lists/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeTwoLists(self, l1: ListNode, l2: ListNode) -> ListNode:
        # use a dummy node for the newHead
        curr = dummyNode = ListNode()
        
        # while both have next
        while l1 and l2:
            # append the smaller node to the combined list
            if l1.val < l2.val:
                curr.next = l1
                l1 = l1.next
            else:
                curr.next = l2
                l2 = l2.next
            
            curr = curr.next
        
        # append the list that has not reached the end to the combined list
        curr.next = l1 if l1 else l2
        
        # get the real head and delete the dummy node
        newHead = dummyNode.next
        del dummyNode
        
        return newHead